### PR TITLE
Iframe the template editor

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

(Also enables the device previews which previously didn't work.)

Why is this important?

For the users, it will make the editor more like the front end. Imagine using `vw` units, a cover background, or responsive fonts, this will now work correctly.
For block authors, it is perfect for easing into iframe context and making sure blocks work there.

Eventually all editors that have theme styles should be iframed, see #21102 for the reasoning, but it's not trivial to implement in the post editor because of backwards compatibility. All block need to continue to work, but with the iframe, some blocks may need to be updated to not use globals and refs instead.

Since the template editor is new, we don't need all blocks to work here (same as FSE) while it also offers us a nice transitioning place for pushing plugins to work with the iframe.

If we don't do this before we release the template editor, we will need to make blocks work flawlessly that have already been saved here before.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by youknowriad** [comment]
I think we should write a dev note about iframing the editors for block authors for the next release. Explain to them why it's important, how it works, how it may break some of their blocks and how to address it.

**Comment by youknowriad** [comment]
It would be good to test this PR properly in order to see whether we can include in 10.6

File: packages/edit-post/src/components/visual-editor/index.js [hunk 1]
```diff
@@ -39,9 +40,44 @@ import { __ } from '@wordpress/i18n';
...
-export default function VisualEditor( { styles } ) {
-	const ref = useRef();
+function MaybeIframe( {
+	children,
+	contentRef,
+	isTemplateMode,
+	styles,
+	style,
+} ) {
+	const ref = useMouseMoveTypingReset();
```
Comment (youknowriad) [inline packages/edit-post/src/components/visual-editor/index.js:51]: What is this hook about?

File: packages/edit-post/src/components/visual-editor/index.js [hunk 2]
```diff
@@ -101,6 +141,19 @@ export default function VisualEditor( { styles } ) {
...
+	// Allow scrolling "through" popovers over the canvas. This is only called
+	// for as long as the pointer is over a popover. Do not use React events
+	// because it will bubble through portals.
+	const toolWrapperRef = useRefEffect( ( node ) => {
+		function onWheel( { deltaX, deltaY } ) {
+			contentRef.current.scrollBy( deltaX, deltaY );
+		}
+		node.addEventListener( 'wheel', onWheel );
+		return () => {
+			node.addEventListener( 'wheel', onWheel );
```
Comment (youknowriad) [inline packages/edit-post/src/components/visual-editor/index.js]: Should this be removeEventListener?

File: packages/edit-post/src/components/visual-editor/index.js [hunk 3]
```diff
@@ -101,6 +141,19 @@ export default function VisualEditor( { styles } ) {
 
 	const blockSelectionClearerRef = useBlockSelectionClearer( true );
 
+	// Allow scrolling "through" popovers over the canvas. This is only called
+	// for as long as the pointer is over a popover. Do not use React events
+	// because it will bubble through portals.
+	const toolWrapperRef = useRefEffect( ( node ) => {
```
Comment (youknowriad) [inline packages/edit-post/src/components/visual-editor/index.js:151]: If feels like this has the potential to break popovers with scrollbars though?

**Comment by ellatrix** [comment]
@youknowriad Yes, this will need a dev note. Worth noting that this is for new (site) content and not for post content, so I think we can feel a bit freer to break backward compatibility with this new feature and try to resolve issues in preparation for iframing the post editor. If we don't do this now, we'll find it a lot hard to get any real world testing.

File: packages/edit-post/src/components/visual-editor/index.js (hunk 3 above)
Comment (ellatrix) [reply packages/edit-post/src/components/visual-editor/index.js:151]: Those are rendered in the normal popover slot, which is outside this div. This slot is solely used for popovers that are placed over the content.

File: packages/edit-post/src/components/visual-editor/index.js (hunk 1 above)
Comment (ellatrix) [reply packages/edit-post/src/components/visual-editor/index.js:51]: It's the same one we have in the site editor. It tracks mouse movement outside the iframe for setting isTyping to false. With #31280, it would be integrated in the block editor so the implementor doesn't need to think about it.

**Comment by jameskoster** [comment]
The toolbars visually breaking out of the frame looks a little strange. It seems there's a balance between usability and visual awkwardness to negotiate there. Maybe it's ok... I'm not sure? :) 

Screenshot 2021-05-04 at 10 57 10

---

I like that the back button is now fixed in position. I do think that the frame can probably sit flush against the bottom of the window in desktop orientation though.

| 👍 | 👎 |
| --- | --- |
| Screenshot 2021-05-04 at 11 01 02 | Screenshot 2021-05-04 at 11 01 20 |

I'm sure it's unrelated, but noting just in case: I seem to recall @youknowriad adding an interaction where clicking on the frame would deselect all blocks. That doesn't seem to be working here.

**Comment by ellatrix** [comment]
Yes, the difference here is that the iframe is creating the scroll area, not the skeleton area, which is just like the front-end.
We can remove the bottom padding if you like.
I'll fix the selection clearer.

**Comment by jasmussen** [comment]
> The toolbars visually breaking out of the frame looks a little strange. It seems there's a balance between usability and visual awkwardness to negotiate there. Maybe it's ok... I'm not sure? :)

I think it's fine for the block UI to be a bit "magical" :)

I'd also agree on removing the bottom padding, so it looks like you'r seeing the top of the document.

As far as dev notes, I'd love for them to truly emphasize the benefits of the iframe and indicate those benefits should ideally reach all editors. Hopefully that would also help us avoid a fragmentation of blocks that work in the post editor sans-iframe, but not the site editor. Accurate media queries and sandboxed CSS in both directions have such value they should be everywhere 🎉

File: packages/block-editor/src/components/block-selection-clearer/index.js [hunk 4]
```diff
@@ -10,22 +10,18 @@ import { useRefEffect } from '@wordpress/compose';
...
 
 export function useBlockSelectionClearer( onlySelfClicks = false ) {
-	const hasSelection = useSelect( ( select ) => {
-		const { hasSelectedBlock, hasMultiSelection } = select(
-			blockEditorStore
-		);
-
-		return hasSelectedBlock() || hasMultiSelection();
-	} );
+	const { hasSelectedBlock, hasMultiSelection } = useSelect(
```
Comment (ntsekouras) [inline packages/block-editor/src/components/block-selection-clearer/index.js]: Why this change was needed?

**Comment by ellatrix** [comment]
> I'd also agree on removing the bottom padding, so it looks like you'r seeing the top of the document.

It would be nice, but then when you're all the way at the bottom, there wouldn't be any padding. I'm not sure how to handle that when the scroll area is within the iframe, but regardless of any fix, the scroll area has to be in the iframe and we can't add the padding inside the iframe. :)

File: packages/block-editor/src/components/block-selection-clearer/index.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-selection-clearer/index.js]: With #31078, any selector calls where the result is used in an event callback should be called only when necessary in the callback instead, and in addition it will always be fresh. Not sure why the component didn't re-render before, it seems worth investigating.

**Comment by jameskoster** [comment]
> It would be nice, but then when you're all the way at the bottom, there wouldn't be any padding.

I think that's ok. It's not really any different to the post editor, no?

**Comment by ellatrix** [comment]
Ok, I removed the padding at the bottom.

**Comment by jameskoster** [comment]
One small hair to split on the design side – now that the bottom padding is gone, the border radii on the bottom of the iframe look a little strange, we should probably set that to 0.

**Comment by ellatrix** [comment]
@jameskoster I removed the border and border radius on the bottom.

**Review by jameskoster — APPROVED** [review]
LGTM on the design side!

**Comment by youknowriad** [comment]
Let's give it a try.

**Comment by FeliciaAKelley** [comment]
25 hidden ... Test and check again ...inside of iframe...from Wordpress..
